pyccel/ast/basic.py

```diff
@@ -23,13 +27,217 @@ def __new__(cls, *args, **kwargs):
+                      The types which we are looking for
+        excluded_nodes : tuple of types
+                      Types for which get_user_nodes should not be called
+
+        Results
+        -------
+        Boolean : True if one of the user nodes is an instance of
+                  the class in the argument
+        """
+        from pyccel.ast.internals import PyccelSymbol
+        excluded_nodes += (PyccelSymbol,)
+        if len(self._user_nodes) == 0:
+            return []
+        else:
+            results  = [p for p in self._user_nodes if isinstance(p, search_type)]
+            results += [r for p in self._user_nodes if not isinstance(p, (search_type, *excluded_nodes)) \
+                    for r in p.get_user_nodes(search_type)]
+            return results
+
+    def get_attribute_nodes(self, search_type, excluded_nodes = ()):
+        """ Returns all objects of the requested type
+        in the current object
+
+        Parameters
+        ----------
+        search_type : ClassType or tuple of ClassTypes
+                      The types which we are looking for
+        excluded_nodes : tuple of types
+                      Types for which get_attribute_nodes should not be called
+
+        Results
+        -------
+        list : List containing all objects of the
+               requested type which exist in self
+        """
+        from pyccel.ast.internals import PyccelSymbol
+        excluded_nodes += (PyccelSymbol,)
+        results = []
+        for n in self._attribute_nodes:
+            v = getattr(self, n)
+            if isinstance(v, search_type):
+                results.append(v)
+
+            if isinstance(v, tuple):
+                for vi in v:
+                    if isinstance(vi, search_type):
+                        results.append(vi)
+
+                    elif vi is not None and not isinstance(vi, excluded_nodes):
+                        results.extend(vi.get_attribute_nodes(search_type))
+
+            elif v is not None and not isinstance(v, excluded_nodes):
+                results.extend(v.get_attribute_nodes(search_type))
+
+        return results
+
+    def substitute(self, original, replacement, excluded_nodes = ()):
+        """
+        Substitute object 'original' for object 'replacement' in the code.
+        Any types in excluded_nodes will not be visited
+
+        Parameters
+        ==========
+        original    : object or tuple of objects
+                      The original object to be replaced
+        replacement : object or tuple of objects
+                      The object which will be inserted instead
+        excluded_nodes : tuple of types
+                      Types for which substitute should not be called
+        """
+        if isinstance(original, tuple):
+            assert(isinstance(replacement, tuple))
+            assert(len(original) == len(replacement))
+        else:
+            original = (original,)
+            substitute = (substitute,)
```

Comment:
substitute where did you initialize this one ?

Reply:
You are right, this is the wrong name. It should be `substitute`->`replacement`
Fixed in 96906ae